### PR TITLE
Telegram bot

### PR DESCRIPTION
**Pull request by toshka**
Hi.
Sometimes it is too lazy to open server console to add new torrent. So I decided to extend telegram bot capabilities. Now it is fully working and configurable but I have few open questions and hope you will help me.
1. What is the best way to run bot on the server? Right now I temporary use screen, but we need some production-ready solution. Or at least describe solutions at README
2. Documentation is not ready. I cant decide where to write it.

### CODE REVIEW

**Comment by idlesign** [comment]
Hi,

Some time ago I was thinking about daemon process, which ought to handle UI (CLI, Web, GTK) commands, maybe this is the way to go. Yet daemon and API needs to be designed. Such approach opens various possible capabilities, including rich UI and bot endpoints.

**Comment by NecroKote** [comment]
Hey folks,
why don't schedule `torrt` with some specific command like `tgbot_pull`, the same way current `walk` implemented? 

(Uponn execution, `torr` will pull new messages from TG and process them.)

Yes, that is not real-time experience bots usually have, but IMHO this way we get consistent experience, similar to current torrt workflow (cron-scheduling).
Also, it is pretty simple, and do not require additional layer of complexity (like residential daemon or API servvice).

**Comment by idlesign** [comment]
Hi,
Sure, that might work with ``getUpdates``, but not webhooks (offering more responsive UX).

**Comment by NecroKote** [comment]
IMHO webhooks, in this case, is a redundant feature.

The difference between adding torrent right away and with the delay of 1-2 min. is not that significant to complicate the existing solution.

File: torrt/bots/telegram_bot.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+        """
+        :param token: str - Telegram's bot token
+        :param allowed_users: str - comma-joined list of users allowed to add new torrents.
+        """
+
+        self.token = token
+        allowed_users = allowed_users or ''
+        self.handler_kwargs = {}
+        try:
+            from telegram.ext import Updater, Filters
```
Comment (idlesign) [inline torrt/bots/telegram_bot.py]: Let's make package import at the top level, like

```
try:
    import telegram
except ImportError:
    telegram = None
```

Than we check for ``None`` in funcs.

File: torrt/bots/telegram_bot.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+            },
+
+            fallbacks=[CommandHandler('cancel', cancel)]
+        )
+
+        self.dispatcher.add_handler(conv_handler)
+        self.dispatcher.add_handler(CommandHandler('add', add_torrent, **self.handler_kwargs))
+
+
+def asking_url_handler(bot, update):
```
Comment (idlesign) [inline torrt/bots/telegram_bot.py]: I'd rather expect those handler functions as methods to not to polute module namespace.

**Comment by idlesign** [comment]
@toshka 
1. Let's settle with ``cron``-based solution for now.
2. I expect here https://github.com/idlesign/torrt/blob/master/docs/source/toolbox.rst and there https://github.com/idlesign/torrt/blob/master/docs/source/commands.rst

@NecroKote Deamon will be implemented someday if I had time %)

**Comment by NecroKote** [comment]
@idlesign, if you'll have some time to describe your vision on 'daemonized' `torrt`, I'll be glad to look into a way of implementing it! Maybe we need new 'Issue' to start discussion ?

**Comment by toshka** [comment]
I thinking a lot about torrt daemon. Right now for me best solution is supervisor (http://supervisord.org/). Because:

- It is has good documentation
- It is easy configurable and don't uses much resources.
- Finally, It is already installed at my server.

May be we just write some README file now?

**Comment by idlesign** [comment]
@NecroKote Some thoughts in #28 

@toshka 
<[Right now for me best solution is supervisor]>
Supervisord is a process control, but I'm talking about daemonization as in #28. Sure you would be able to use Supervisord or Upstart or Systemd, etc. %)

<[May be we just write some README file now?]>
Sure, we've already settled that the issue won't depend on daemon to be merged. You can update README with bot usage examples, but I'd rather expect them to be in the documentaion. A step by step description of how to configure and use bot commands would be very nice.

File: torrt/base_bot.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
+from utils import WithSettings, BotObjectsRegistry
+
+
+class BaseBot(WithSettings):
```
Comment (idlesign) [inline torrt/base_bot.py]: Maybe drop ``bot`` in favor of ``actor`` (as in #28) or something else more generic to describe subsystem built to reach out and communicate with remote systems, e.g. messengers? Becase only ``bot`` seems too narrow.
How do you think?

Comment (toshka) [reply torrt/base_bot.py]: I think we can do it later when all became clear with #28

Comment (idlesign) [reply torrt/base_bot.py]: #28 here is just for illustration.

Unfortunatelly we won't be able to change anything after the release (otherwise we break the backward compatibility — both API and CLI), and would be obliged to use e.g. ``configure_bot`` (instead of, say ``configure_actor``) forever even its'possible that it'd be semantically incorrect to name other integrations as ``bots``. We certainly could initiate some deprectation processes but that'd be painful both for us and for end-users. That's why I propose a more generic name.

File: docs/source/bot.rst [hunk 4]
```diff
@@ -0,0 +1,71 @@
+Telegram Bot
+============
+
+You can add new torrents with Telegram bot. Here described how to configure and run it.
```
Comment (idlesign) [inline docs/source/bot.rst]: We need to mention what Telegram library we use. Maybe even add that into ``extras`` section of ``setup.py``.

File: docs/source/bot.rst [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+
+1. Install supervisor on your host as described at http://supervisord.org/installing.html
+2. Create configuration file torrt.conf at */etc/supervisor/conf.d/* with following content::
+
+    [program:torrt]
+    directory=/tmp
+    command=PATH_TO_TORRT_SCRIPT run_bots
+    user=USER_ON_HOST
+    autostart=true
+    autorestart=true
```
Comment (idlesign) [inline docs/source/bot.rst]: I couldn't find the library we use, so I wonder, wether the Updater runs a loop, or you just constantly restarting ``torrt`` with supervisor?

Comment (NecroKote) [reply docs/source/bot.rst]: From the code the package is `python-telegram-bot` .
It runs a loop, so after `torrt run_bots` the process stays up and processes incoming events.

Comment (toshka) [reply docs/source/bot.rst]: Yes, @NecroKote absolutely right!

File: torrt/base_bot.py (hunk 3 above)
Comment (toshka) [reply torrt/base_bot.py]: It is easy to do, you know. But seems for usual user it is hard to understand what is actor. Generally even I can't imagine another way to communicate with actor. I know that it can appears in the future and we should be ready for it. So I can rename bots to actors if you want. But I need help with documentation in this case :)

File: setup.py [hunk 6]
```diff
@@ -48,6 +48,9 @@ def get_version():
     ],
     setup_requires=[] + (['pytest-runner'] if 'test' in sys.argv else []) + [],
 
+    extras_require={
+        'telegram': ['python-telegram-bot']
```
Comment (idlesign) [inline setup.py]: Do we need to pin down the lowest version number?

File: torrt/base_bot.py (hunk 3 above)
Comment (idlesign) [reply torrt/base_bot.py]: Ok, lt's settle with ``bot``.

File: setup.py (hunk 6 above)
Comment (toshka) [reply setup.py]: Yes, but more important was to pin higher version because maintainers are working on major release version 12 with lot of changes.

File: torrt/notifiers/telegram.py [hunk 7]
```diff
@@ -32,11 +32,13 @@ def test_configuration(self):
 
     def send_message(self, msg):
         url = '%s%s/sendMessage' % (self.url, self.token)
-        r = requests.post(url, data={'chat_id': self.chat_id, 'text': msg})
-        if r.json()['ok']:
-            LOGGER.info('Telegram message was sent to user %s' % self.chat_id)
-        else:
-            LOGGER.error('Telegram notification not send: %s' % r['description'])
+        response = requests.post(url, data={'chat_id': self.chat_id, 'text': msg})
```
Comment (idlesign) [inline torrt/notifiers/telegram.py]: Please do not introduce unrelated changes into the pull request.

File: setup.py (hunk 6 above)
Comment (idlesign) [reply setup.py]: Are they already guilty in backward incompatible releases? %)

Comment (toshka) [reply setup.py]: I saw a lot of warnings in console for beta version so decide exclude 12 version for now.

File: torrt/notifiers/telegram.py (hunk 7 above)
Comment (toshka) [reply torrt/notifiers/telegram.py]: Of cause, sorry

**Comment by idlesign** [comment]
Thank you. Merged.
